<h1> H1B Reports 2024 </h1>

In [2]:
# reading the document in CSV
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
employer = pd.read_csv("employer.csv", encoding='utf-16', sep='\t', index_col=0)
employer.head()

Fiscal Year                    Employer (Petitioner) Name  \
Line by line                                                              
1                       2024                                        NaN   
2                       2024                                        NaN   
3                       2024                      1 800 FLOWERS COM INC   
4                       2024  1 HOTEL KAUAI LLC DBA 1 HOTEL HANALEI BAY   
5                       2024                            1 OAK MEDIA LLC   

              Tax ID                              Industry (NAICS) Code  \
Line by line                                                              
1             3581.0  54 - Professional, Scientific, and Technical S...   
2             4245.0             62 - Health Care and Social Assistance   
3             7311.0                               44-45 - Retail Trade   
4             5669.0               72 - Accommodation and Food Services   
5             7999.0  54 - Professional, Scientific, and Technical S...   

             Petitioner City Petitioner State  Petitioner Zip Code  \
Line by line                                                         
1                  LAFAYETTE               CA              94549.0   
2                      DAVIE               FL              33328.0   
3                    JERICHO               NY              11753.0   
4                      MIAMI               FL              33133.0   
5               LINCOLN CITY               OR              97367.0   

             Initial Approval  Initial Denial Continuing Approval  \
Line by line                                                        
1                           0               0                   1   
2                           1               0                   0   
3                           1               0                   5   
4                           0               0                   1   
5                           1               0                   0   

              Continuing Denial  
Line by line                     
1                             0  
2                             0  
3                             0  
4                             0  
5                             0

In [4]:
# Checking for missing values
employer.isnull().sum()

Fiscal Year                      0
Employer (Petitioner) Name       2
Tax ID                         417
Industry (NAICS) Code         2375
Petitioner City                  0
Petitioner State                 4
Petitioner Zip Code              7
Initial Approval                 0
Initial Denial                   0
Continuing Approval              0
Continuing Denial                0
dtype: int64

In [5]:
# Addressing `line by line` column 
employer = employer.reset_index()
employer.head()

Line by line  Fiscal Year                    Employer (Petitioner) Name  \
0            1            2024                                        NaN   
1            2            2024                                        NaN   
2            3            2024                      1 800 FLOWERS COM INC   
3            4            2024  1 HOTEL KAUAI LLC DBA 1 HOTEL HANALEI BAY   
4            5            2024                            1 OAK MEDIA LLC   

   Tax ID                              Industry (NAICS) Code Petitioner City  \
0  3581.0  54 - Professional, Scientific, and Technical S...       LAFAYETTE   
1  4245.0             62 - Health Care and Social Assistance           DAVIE   
2  7311.0                               44-45 - Retail Trade         JERICHO   
3  5669.0               72 - Accommodation and Food Services           MIAMI   
4  7999.0  54 - Professional, Scientific, and Technical S...    LINCOLN CITY   

  Petitioner State  Petitioner Zip Code Initial Approval  Initial Denial  \
0               CA              94549.0                0               0   
1               FL              33328.0                1               0   
2               NY              11753.0                1               0   
3               FL              33133.0                0               0   
4               OR              97367.0                1               0   

  Continuing Approval  Continuing Denial  
0                   1                  0  
1                   0                  0  
2                   5                  0  
3                   1                  0  
4                   0                  0

In [6]:
employer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61456 entries, 0 to 61455
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Line by line                61456 non-null  object 
 1   Fiscal Year                 61456 non-null  int64  
 2   Employer (Petitioner) Name  61454 non-null  object 
 3   Tax ID                      61039 non-null  float64
 4   Industry (NAICS) Code       59081 non-null  object 
 5   Petitioner City             61456 non-null  object 
 6   Petitioner State            61452 non-null  object 
 7   Petitioner Zip Code         61449 non-null  float64
 8   Initial Approval            61456 non-null  object 
 9   Initial Denial              61456 non-null  int64  
 10  Continuing Approval         61456 non-null  object 
 11  Continuing Denial           61456 non-null  int64  
dtypes: float64(2), int64(3), object(7)
memory usage: 5.6+ MB


In [7]:
print(employer.columns)

Index(['Line by line', 'Fiscal Year   ', 'Employer (Petitioner) Name',
       'Tax ID', 'Industry (NAICS) Code', 'Petitioner City',
       'Petitioner State', 'Petitioner Zip Code', 'Initial Approval',
       'Initial Denial', 'Continuing Approval', 'Continuing Denial'],
      dtype='object')


<p><b> we are going to drop Fiscal Year, Tax iD, Petitioner Zip Code</b> </p>

In [8]:
employer = employer.drop(['Line by line','Fiscal Year   ','Tax ID','Petitioner Zip Code'],axis=1)

In [9]:
employer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61456 entries, 0 to 61455
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Employer (Petitioner) Name  61454 non-null  object
 1   Industry (NAICS) Code       59081 non-null  object
 2   Petitioner City             61456 non-null  object
 3   Petitioner State            61452 non-null  object
 4   Initial Approval            61456 non-null  object
 5   Initial Denial              61456 non-null  int64 
 6   Continuing Approval         61456 non-null  object
 7   Continuing Denial           61456 non-null  int64 
dtypes: int64(2), object(6)
memory usage: 3.8+ MB


In [10]:
employer.isnull().sum()

Employer (Petitioner) Name       2
Industry (NAICS) Code         2375
Petitioner City                  0
Petitioner State                 4
Initial Approval                 0
Initial Denial                   0
Continuing Approval              0
Continuing Denial                0
dtype: int64

In [11]:
employer['Industry (NAICS) Code'].unique()

array(['54 - Professional, Scientific, and Technical Services',
       '62 - Health Care and Social Assistance', '44-45 - Retail Trade',
       '72 - Accommodation and Food Services',
       '52 - Finance and Insurance', '51 - Information',
       '31-33 - Manufacturing',
       '81 - Other Services (except Public Administration)',
       '42 - Wholesale Trade', nan,
       '53 - Real Estate and Rental and Leasing',
       '56 - Administrative and Support and Waste Management and Remediation Services',
       '23 - Construction', '61 - Educational Services',
       '71 - Arts, Entertainment, and Recreation',
       '92 - Public Administration',
       '11 - Agriculture, Forestry, Fishing and Hunting',
       '48-49 - Transportation and Warehousing',
       '55 - Management of Companies and Enterprises', '22 - Utilities',
       '21 - Mining, Quarrying, and Oil and Gas Extraction'], dtype=object)

In [12]:
employer['Employer (Petitioner) Name'].nunique()

52892

<p><b> We have a high number of missing values in `Industry (NAICS) Code` (2375). We need to address this high number of missing values</b></p>
<li>We can try imputation based on a placeholder -- we can keep track of missing values  using a special form of 'unkown'</li>
<li> Imputation based on another column -- We can check the   `employer name` and try to fit or associate with specific industries</li>
<p> Handling the missing values we replace the missing values with 'Unknown' category as per now </p> 

In [13]:
employer['Industry (NAICS) Code'] = employer['Industry (NAICS) Code'].fillna('unknown')

<p><b>Creating a visualization to check the number of occurences per each 'NAICS code'</b></p>

In [14]:
# PLotly implementation
import plotly.express as px
from IPython.display import display
import webbrowser


naics_counts = employer['Industry (NAICS) Code'].value_counts().reset_index()
naics_counts.columns = ['NAICS Code', 'Number of Occurrences']

fig = px.bar(naics_counts, 
             x='NAICS Code', 
             y='Number of Occurrences', 
             title='NAICS Code Distribution',
             color='Number of Occurrences', 
             color_continuous_scale='viridis')

fig.update_layout(xaxis_title='NAICS Code',
                  yaxis_title='Number of Occurrences',
                  xaxis_tickangle=-90,
                  height=1000,
                  width=1000)

# fig.show()
# display(fig)
fig.write_html("plot.html")
webbrowser.open("plot.html")

True

<p> The unkown values are still high so we will need to check on this for future recommendations </p>
<p> 'Employer (Petitioner) Name' has only 2 missing values. These might not necessarily affect the interpretation of the data so we will drop the 2 values. </p>

In [15]:
employer.shape

(61456, 8)

In [16]:
# employer['Employer (Petitioner) Name', 'Petitioner State'].dropna(inplace=True)
employer.dropna(subset=['Employer (Petitioner) Name', 'Petitioner State'], inplace=True)

In [17]:
employer.isnull().sum()

Employer (Petitioner) Name    0
Industry (NAICS) Code         0
Petitioner City               0
Petitioner State              0
Initial Approval              0
Initial Denial                0
Continuing Approval           0
Continuing Denial             0
dtype: int64

In [18]:
employer.shape

(61450, 8)

In [19]:
employer.head(10)

Employer (Petitioner) Name  \
2                       1 800 FLOWERS COM INC   
3   1 HOTEL KAUAI LLC DBA 1 HOTEL HANALEI BAY   
4                             1 OAK MEDIA LLC   
5                          1 POINT SYSTEM LLC   
6                         1 REALTY GLOBAL INC   
7                               1 WAYDOOR LLC   
8                   1CONNECT TECHNOLOGIES INC   
9                      1LIFE HEALTHCARE, INC.   
10                           1POINT3ACRES LLC   
11                                 1RIVET LLC   

                                Industry (NAICS) Code Petitioner City  \
2                                44-45 - Retail Trade         JERICHO   
3                72 - Accommodation and Food Services           MIAMI   
4   54 - Professional, Scientific, and Technical S...    LINCOLN CITY   
5   54 - Professional, Scientific, and Technical S...       FORT MILL   
6                          52 - Finance and Insurance          WALNUT   
7   54 - Professional, Scientific, and Technical S...         CUMMING   
8   54 - Professional, Scientific, and Technical S...      ALPHARETTA   
9              62 - Health Care and Social Assistance   SAN FRANCISCO   
10                                   51 - Information        BELLEVUE   
11  54 - Professional, Scientific, and Technical S...          RESTON   

   Petitioner State Initial Approval  Initial Denial Continuing Approval  \
2                NY                1               0                   5   
3                FL                0               0                   1   
4                OR                1               0                   0   
5                SC                1               0                   1   
6                CA                1               0                   0   
7                GA                0               0                   1   
8                GA                0               1                   3   
9                CA                0               0                   1   
10               WA                1               0                   0   
11               VA                0               0                   1   

    Continuing Denial  
2                   0  
3                   0  
4                   0  
5                   0  
6                   0  
7                   0  
8                   0  
9                   0  
10                  0  
11                  0

<p> We have only dropped 6 rows, this does not necessarily affect the interpretation of the data. </p>

<h5><b> 1. Finding the distribution of VISAs in different states </b></h5>

In [20]:
import plotly.express as px


state_counts = employer['Petitioner State'].value_counts().reset_index()
state_counts.columns = ['State', 'Number of Occurrences']

fig = px.choropleth(
    state_counts,
    locations='State',
    locationmode='USA-states',
    color='Number of Occurrences',
    scope='usa',
    color_continuous_scale='viridis',
    title='H1B Distributions by State'
)

fig.update_layout(
    width=1200, 
    height=800   
)
fig.show()



In [30]:
# import plotly.express as px

state_counts = employer['Petitioner State'].value_counts().reset_index() 
state_counts.columns = ['State', 'Number of Occurrences']

fig = px.bar(
    state_counts,
    x='State',
    y='Number of Occurrences',
    title='H1B Distributions by State',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)

fig.update_layout(
    width=1200,
    height=800,
    xaxis_title='State',
    yaxis_title='Number of H1B Applications',
    xaxis={'categoryorder': 'total descending'}
)

fig.show()

<p><b>California</b> appears to have the <b>highest number of H-1B</b> visa occurrences, shown in bright yellow with numbers appearing to be around 10,000 or higher </p>

<p>Texas shows a significant number as well, appearing in green, suggesting approximately 6,000-7,000 occurrences.
New York also has a notable concentration, shown in a teal/green color.</p>

<p><b>Most states in the Midwest, Mountain West, and parts of the South</b> appear in dark purple, <b>indicating very low numbers of H-1B visas.</b></p>

<p><b>Florida and Illinois</b> show moderate numbers in blue tones, suggesting they have more H-1B visas than most states but significantly fewer than California, Texas, and New York.
The Northeast corridor has varying levels, with some states showing higher concentrations than others.</p>

<p>This distribution aligns with what we might expect, as H-1B visas are typically associated with technology, research, and other specialized industries that tend to be concentrated in states with large tech hubs like California's Silicon Valley, Texas's tech corridors, and New York's business centers.</p>

<h5><b> Filtering by industry for the top 25 states</b><h5>
<p> We need to figure the industries in the top 25 states with the most sponsorship</p>

In [ ]:

# Get top 25 states by petition count
state_counts = employer['Petitioner State'].value_counts().reset_index() 
state_counts.columns = ['State', 'Number of Occurrences']
top_25_states = state_counts.head(25)['State'].tolist()

# Filter data for only top 25 states
top_states_data = employer[employer['Petitioner State'].isin(top_25_states)]

# Get industry distribution in top 25 states
industry_counts = top_states_data['Industry (NAICS) Code'].value_counts().reset_index()
industry_counts.columns = ['Industry', 'Number of Occurrences']

# Create bar chart for industries
fig = px.bar(
    industry_counts.head(20),  # Show top 20 industries
    x='Industry',
    y='Number of Occurrences',
    title='Top Industries in Top 25 States by H1B Petitions',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)

fig.update_layout(
    width=1400,
    height=800,
    xaxis_title='Industry',
    yaxis_title='Number of H1B Petitions',
    xaxis={'categoryorder': 'total descending'}
)
fig.update_xaxes(tickangle=45)

fig.show()

print("Top 25 States:")
print(top_25_states)
print(f"\nTotal petitions in top 25 states: {len(top_states_data)}")
print(f"Total petitions overall: {len(employer)}")
print(f"Percentage covered by top 25 states: {len(top_states_data)/len(employer)*100:.1f}%")

# Create a pivot table to see industry distribution across top states
state_industry_pivot = top_states_data.groupby(['Petitioner State', 'Industry (NAICS) Code']).size().reset_index(name='Count')
state_industry_summary = state_industry_pivot.pivot_table(
    index='Petitioner State', 
    columns='Industry (NAICS) Code', 
    values='Count', 
    fill_value=0
)

print("\nTop 5 industries by state (sample):")
print(state_industry_summary.iloc[:5, :5])  


Top 25 States:
['CA', 'TX', 'NY', 'NJ', 'IL', 'MA', 'FL', 'GA', 'VA', 'MI', 'PA', 'OH', 'NC', 'MD', 'WA', 'AZ', 'MN', 'MO', 'CO', 'CT', 'TN', 'IN', 'WI', 'DC', 'UT']

Total petitions in top 25 states: 56075
Total petitions overall: 61450
Percentage covered by top 25 states: 91.3%

Top 5 industries by state (sample):
Industry (NAICS) Code  11 - Agriculture, Forestry, Fishing and Hunting  \
Petitioner State                                                         
AZ                                                                 2.0   
CA                                                                30.0   
CO                                                                 2.0   
CT                                                                 0.0   
DC                                                                 0.0   

Industry (NAICS) Code  21 - Mining, Quarrying, and Oil and Gas Extraction  \
Petitioner State                                                            
AZ       

<h5><b> State vs City Distribution </b></h5>

In [21]:
# city_counts = employer['Petitioner City'].value_counts().reset_index()
city_state = employer.groupby(['Petitioner City', 'Petitioner State']).size().reset_index(name="Number of Occurrences")
# city_state = city_state.sort_values(by='Number of Occurences', ascending=False) 

#  Get the top 20 cities and States
top_cities = city_state.sort_values('Number of Occurrences', ascending=False).head(40)

fig = px.treemap(
    top_cities,
    path=['Petitioner State', 'Petitioner City'],
    values='Number of Occurrences',
    title='H1B Petitions by State and City',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)

fig.update_layout(
    width=1200,
    height=800
)

fig.show()

<p> New York has the highest number of approvals within cities. This shows that NY is a good city for international students </p>

<h5><b> 3. Employer Distribution across Counts </b></h5>

In [22]:
petitioner = employer['Employer (Petitioner) Name'].value_counts().reset_index()
petitioner.columns = ['Employer Name', 'Number of Occurrences']

fig = px.bar(
    petitioner.head(50),
    x='Employer Name',
    y='Number of Occurrences',
    title='Top 20 Employers by Number of Petitions',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)

fig.update_layout(
    width=1200,
    height=800
)
fig.update_xaxes(tickangle=45)

fig.show()

<p> The <b>State University of New York </b> registered the highest numbers of H1B applications </p> 

<h5><b> 4. Grouping the NAICS code </b></h5>
<p> This is to find the industries with the highest H1B applications </p>

In [23]:
naics = employer['Industry (NAICS) Code'].value_counts().reset_index()
naics.columns = ['NAICS Code', 'Number of Occurrences']

# fig = px.bar(
#     naics.head(50),
#     x='NAICS Code',
#     y='Number of Occurrences',
#     title='Top 20 NAICS Codes by Number of Petitions',
#     color='Number of Occurrences',
#     color_continuous_scale='viridis'
# )
fig = px.sunburst(
    naics.head(10),
    path=['NAICS Code'],
    values='Number of Occurrences',
    title='Top 20 Employers by Number of Petitions',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)
fig.update_layout(
    width=1200,
    height=800
)

fig.show()

<h5><b> 5. Employer Vs State Counts </b></h5>  

In [24]:
emp_grp = employer.groupby(['Industry (NAICS) Code', 'Petitioner State',]).size().reset_index(name="Number of Occurrences")
top_emp_grp = emp_grp.sort_values(by='Number of Occurrences', ascending=False)
# emp_grp.head()

fig = px.treemap(
    top_emp_grp.head(20),
    path=['Petitioner State', 'Industry (NAICS) Code'],
    values='Number of Occurrences',
    title='Grouping H1B Petitions by State and Industry',
    color='Number of Occurrences',
    color_continuous_scale='viridis'
)
fig.update_layout(
    width=1200,
    height=800
) 
fig.show()

<p> In most cases the <b> Professional, Scientific, and Technical Services </b> seems to have the most H1B applications </b></p>

<h5><b> 6. Approval Rates (Initial & Continuing) Vs Industry (NAICS) code </b></h5>

In [25]:

approvals = employer.groupby(['Initial Approval', 'Continuing Approval','Industry (NAICS) Code']).sum().reset_index() 

approvals_melted = approvals.melt(id_vars='Industry (NAICS) Code', 
                                  var_name='Approval Type', 
                                  value_name='Total Approvals')

fig = px.bar(approvals_melted,
             x='Industry (NAICS) Code', 
             y='Total Approvals', 
             color='Approval Type', 
             color_continuous_scale='viridis',
             title='Industry Approvals: Initial and Continuing',
             barmode='group',
             color_discrete_sequence=px.colors.qualitative.Vivid)



fig.update_layout(
    width=1200,
    height=800
)
fig.update_xaxes(tickangle=45)

fig.update_layout(
    width=1200,
    height=800
) 

fig.show()

<b>Tasks</b>
1. Group into approvals + Industry
2. Group into denials + Industry
3. Group into approvals + Employers
4. Group into denials + Employers
5. Filter by MI
6. Classify approvals by state + Industry + Employer
7. Classify denials by state + Industry + Employer

<h1> PERM Applications</h1>
<p> Dtata sourced from https://www.dol.gov/agencies/eta/foreign-labor/performance?utm_source=chatgpt.com</p>

In [26]:
# Loading filr in excel
perm = pd.read_excel("PERM.xlsx", sheet_name='PERM_Disclosure_FY2024_Q4',index_col=0)
perm.head()

CASE_STATUS RECEIVED_DATE DECISION_DATE REFILE  \
CASE_NUMBER                                                           
A-23043-00641          Withdrawn    2023-02-12    2023-10-01      N   
A-22256-20380  Certified-Expired    2022-09-22    2023-10-02      N   
A-22271-30061  Certified-Expired    2022-10-24    2023-10-02      N   
A-22272-30740  Certified-Expired    2022-10-24    2023-10-02      N   
A-22185-82932  Certified-Expired    2022-07-26    2023-10-02      N   

              ORIG_FILE_DATE PREVIOUS_SWA_CASE_NUMBER_STATE  \
CASE_NUMBER                                                   
A-23043-00641            NaT                            NaN   
A-22256-20380            NaT                            NaN   
A-22271-30061            NaT                            NaN   
A-22272-30740            NaT                            NaN   
A-22185-82932            NaT                            NaN   

              SCHD_A_SHEEPHERDER                    EMPLOYER_NAME  \
CASE_NUMBER                                                         
A-23043-00641                  N  LIBERTY CIVIL CONSTRUCTION, LLC   
A-22256-20380                  N              Siser North America   
A-22271-30061                  N                     Blend360 LLC   
A-22272-30740                  N                     Blend360 LLC   
A-22185-82932                  N                      AMDOCS INC.   

                       EMPLOYER_ADDRESS_1 EMPLOYER_ADDRESS_2  ...  \
CASE_NUMBER                                                   ...   
A-23043-00641               2937 HERO WAY                NaN  ...   
A-22256-20380             12900 Hall Road          Suite 270  ...   
A-22271-30061       10221 Wincopin Circle          Suite 300  ...   
A-22272-30740       10221 Wincopin Circle          Suite 300  ...   
A-22185-82932  625 MARYVILLE CENTRE DRIVE            STE 200  ...   

              FOREIGN_WORKER_ALT_OCC_EXP FOREIGN_WORKER_EXP_WITH_EMPL  \
CASE_NUMBER                                                             
A-23043-00641                        NaN                            Y   
A-22256-20380                          Y                            N   
A-22271-30061                          Y                            N   
A-22272-30740                          Y                            N   
A-22185-82932                          Y                            N   

              FOREIGN_WORKER_EMPL_PAY_FOR_ED FOREIGN_WORKER_CURR_EMPLOYED  \
CASE_NUMBER                                                                 
A-23043-00641                              Y                            N   
A-22256-20380                              N                            Y   
A-22271-30061                              N                            Y   
A-22272-30740                              N                            Y   
A-22185-82932                              N                            Y   

              EMPLOYER_COMPLETED_APPLICATION     PREPARER_NAME  \
CASE_NUMBER                                                      
A-23043-00641                              Y                     
A-22256-20380                              N   Marshal E Hyman   
A-22271-30061                              N      Sufen  Zhang   
A-22272-30740                              N      Sufen  Zhang   
A-22185-82932                              N  Warren M Schultz   

               PREPARER_TITLE               PREPARER_EMAIL  \
CASE_NUMBER                                                  
A-23043-00641             NaN                          NaN   
A-22256-20380        Attorney     sharonw@marshalhyman.com   
A-22271-30061       Principal  szhang@milesstockbridge.com   
A-22272-30740       Principal  szhang@milesstockbridge.com   
A-22185-82932        Attorney     teamEBUcorp@fragomen.com   

                      EMP_INFO_DECL_NAME  \
CASE_NUMBER                                
A-23043-00641  Brandon F Menjivar Molina   
A-22256-20380            Daniel

In [27]:
# Checking for missing values(Perm Appplication)
perm.isnull().sum()

CASE_STATUS               0
RECEIVED_DATE             0
DECISION_DATE             0
REFILE                    1
ORIG_FILE_DATE        92208
                      ...  
PREPARER_NAME             0
PREPARER_TITLE        10285
PREPARER_EMAIL        10304
EMP_INFO_DECL_NAME        0
EMP_DECL_TITLE            7
Length: 154, dtype: int64

In [28]:
print(employer.shape)
print(employer.columns)

(61450, 8)
Index(['Employer (Petitioner) Name', 'Industry (NAICS) Code',
       'Petitioner City', 'Petitioner State', 'Initial Approval',
       'Initial Denial', 'Continuing Approval', 'Continuing Denial'],
      dtype='object')


In [29]:
# Unique values in the column
employer['Fiscal Year   '].unique()

KeyError: 'Fiscal Year   '